src/easyscience/base_classes/new_base.py

```diff
@@ -0,0 +1,145 @@
+
+    @property
+    def display_name(self) -> str:
+        """
+        Get a pretty display name.
+
+        :return: The pretty display name.
+        """
+        display_name = self._display_name
+        if display_name is None:
+            display_name = self.unique_name
+        return display_name
+
+    @display_name.setter
+    @property_stack
+    def display_name(self, name: str) -> None:
+        """
+        Set the pretty display name.
+
+        :param name: Pretty display name of the object.
+        """
+        if name is not None and not isinstance(name, str):
+            raise TypeError('Display name must be a string or None')
+        self._display_name = name
+
+    def to_dict(self, skip: Optional[List[str]] = None) -> Dict[str, Any]:
+        """
+        Convert an EasyScience object into a full dictionary using `SerializerBase`s generic `convert_to_dict` method.
+
+        :param skip: List of field names as strings to skip when forming the dictionary
+        :return: encoded object containing all information to reform an EasyScience object.
+        """
+        serializer = SerializerBase()
+        if skip is None:
+            skip = []
+        if self._default_unique_name and 'unique_name' not in skip:
+            skip.append('unique_name')
+        if self._display_name is None:
+            skip.append('display_name')
+        return serializer._convert_to_dict(self, skip=skip, full_encode=False)
+
+    @classmethod
+    def from_dict(cls, obj_dict: Dict[str, Any]) -> None:
+        """
+        Re-create an EasyScience object from a full encoded dictionary.
+
+        :param obj_dict: dictionary containing the serialized contents (from `SerializerDict`) of an EasyScience object
+        :return: Reformed EasyScience object
+        """
+        if not SerializerBase._is_serialized_easyscience_object(obj_dict):
+            raise ValueError('Input must be a dictionary representing an EasyScience object.')
+        if obj_dict['@class'] == cls.__name__:
+            kwargs = SerializerBase.deserialize_dict(obj_dict)
+            return cls(**kwargs)
+        else:
+            raise ValueError(f'Class name in dictionary does not match the expected class: {cls.__name__}.')
+
+    def __dir__(self) -> Iterable[str]:
+        """
+        This creates auto-completion and helps out in iPython notebooks.
+
+        :return: list of function and parameter names for auto-completion
+        """
+        new_class_objs = list(k for k in dir(self.__class__) if not k.startswith('_'))
+        return sorted(new_class_objs)
+
+    def __copy__(self) -> NewBase:
+        """Return a copy of the object."""
+        temp = self.to_dict(skip=['unique_name'])
+        new_obj = self.__class__.from_dict(temp)
+        return new_obj
+
+    def __deepcopy__(self, memo):
+        return self.__copy__()
+
+    def __repr__(self) -> str:
+        return f'{self.__class__.__name__} `{self.unique_name}`'
```

Comment:
Will we not need `__reduce__` for the base class? This might come in handy at some point.
(Yes, this is asking for overengineering of options we don't strictly need...)